# Preprocessing

In this notebook, I'll be preprocessing all my data to get it ready to be put into a model.

In [1]:
import pandas as pd
pd.set_option('display.max_columns', 100)
import numpy as np


## Data Loading

Just checking to see if my data has been correctly imported.

In [2]:
df=pd.read_csv("data1/my_dataset.csv")

In [3]:
df.sample(10)

player_id  last_season_x  most_recent_club_id country_of_citizenship  \
39523      197751           2021                 1390                  Italy   
4903       406675           2022                  868                 Brazil   
11204      264423           2021                  206                Denmark   
118010     340172           2020                  403                Albania   
5903       724535           2022                  987                Ireland   
94496      255942           2023                   46                  Italy   
54953      182079           2023                 3209                 Turkey   
13946       44699           2022                 2995               Portugal   
72706      442703           2023                 1005                Albania   
76037       80444           2022                  405                 Brazil   

              sub_position  position   foot  height_in_cm  market_value  \
39523       Centre-Forward    Attack  right         194.0       3000000   
4903             Left-Back  Defender   left         178.0       1200000   
11204       Centre-Forward    Attack  right         188.0       2000000   
118010    Central Midfield  Midfield  right         180.0        650000   
5903          Right Winger    Attack   left         178.0         25000   
94496     Central Midfield  Midfield  right         175.0        500000   
54953     Central Midfield  Midfield  right         179.0        400000   
13946            Left-Back  Defender   left         175.0        300000   
72706   Defensive Midfield  Midfield  right         185.0        450000   
76037          Left Winger    Attack  right         172.0     100000000   

        highest_ever_market_value        date  age  contract_days_left  month  \
39523                   7500000.0  2020-08-25   24                -134      8   
4903                    3800000.0  2020-04-08   23                 416      4   
11204                   4000000.0  2021-06-04   27                 416      6   
118010                  1800000.0  2021-03-09   22                 232      3   
5903                      50000.0  2022-06-27   19                -164      6   
94496                  75000000.0  2015-01-07   18                 962      1   
54953                    675000.0  2020-02-04   29                 597      2   
13946                   6000000.0  2022-05-31   35                -134      5   
72706                    800000.0  2022-06-07   26                 962      6   
76037                 150000000.0  2019-02-28   27                 962      2   

        year  goals  assists  yellow_cards  red_cards  minutes_played  \
39523   2020      2        1             1          0             705   
4903    2020      1        2             0          1            1907   
11204   2021      6        1             0          0             722   
118010  2021      0        0             3          0             455   
5903    2022      0        0             0          0               5   
94496   2015      0        0             2          0             220   
54953   2020      0        1             1          1             913   
13946   2022      4        6            10          0            2582   
72706   2022      1        1             5          0            1702   
76037   2019     10        8             1          0            2882   

        player_club_id league_id  net_transfer_record  national_team_players  
39523             1390       IT1           -8600000.0                      9  
4903              2503       PO1           -2600000.0                      4  
11204              206       DK1            3610000.0                      4  
118010             317       NL1            7100000.0                      3  
5903               987       SC1             586000.0                      4  
94496             1390       IT1           -8600000.0                      9  
54953            39722       TR1              70000.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123594 entries, 0 to 123593
Data columns (total 24 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   player_id                  123594 non-null  int64  
 1   last_season_x              123594 non-null  int64  
 2   most_recent_club_id        123594 non-null  int64  
 3   country_of_citizenship     123594 non-null  object 
 4   sub_position               123594 non-null  object 
 5   position                   123594 non-null  object 
 6   foot                       123594 non-null  object 
 7   height_in_cm               123594 non-null  float64
 8   market_value               123594 non-null  int64  
 9   highest_ever_market_value  123594 non-null  float64
 10  date                       123594 non-null  object 
 11  age                        123594 non-null  int64  
 12  contract_days_left         123594 non-null  int64  
 13  month                      12

- The `player_id` collumn seems relatively useless in the case of predicting market value as its just an identifier.
- The `most_recent_club_id` collumn also seems useless as I already have the player_club_id collumn making it useless.
- We can also drop the `highest_ever_market_value` collumn as its way too correlated (see EDA notebook) with the target variable.

We can drop them both.

In [5]:
df=df.drop(["player_id","most_recent_club_id"],axis=1)

In [6]:
df.describe()

last_season_x   height_in_cm  market_value  highest_ever_market_value  \
count  123594.000000  123594.000000  1.235940e+05               1.235940e+05   
mean     2021.424948     182.265814  5.106035e+06               1.030578e+07   
std         2.298380       6.695002  1.061903e+07               1.747416e+07   
min      2012.000000     159.000000  1.000000e+04               2.500000e+04   
25%      2021.000000     178.000000  5.000000e+05               1.000000e+06   
50%      2023.000000     183.000000  1.500000e+06               3.500000e+06   
75%      2023.000000     187.000000  4.500000e+06               1.200000e+07   
max      2023.000000     207.000000  2.000000e+08               2.000000e+08   

                 age  contract_days_left          month           year  \
count  123594.000000       123594.000000  123594.000000  123594.000000   
mean       25.031830          443.834749       6.428346    2018.691587   
std         3.985102          471.763162       3.587796       2.976855   
min        15.000000         -314.000000       1.000000    2012.000000   
25%        22.000000          232.000000       3.000000    2017.000000   
50%        25.000000          232.000000       6.000000    2019.000000   
75%        28.000000          597.000000      10.000000    2021.000000   
max        44.000000         3154.000000      12.000000    2023.000000   

               goals        assists   yellow_cards      red_cards  \
count  123594.000000  123594.000000  123594.000000  123594.000000   
mean        1.973793       1.501165       2.737706       0.068798   
std         3.695028       2.450115       2.855257       0.265803   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         1.000000       1.000000       2.000000       0.000000   
75%         2.000000       2.000000       4.000000       0.000000   
max        59.000000      33.000000      23.000000       3.000000   

       minutes_played  player_club_id  net_transfer_record  \
count   123594.000000   123594.000000         1.235940e+05   
mean      1308.635727     3175.928945         2.138397e+05   
std       1010.055198     7846.777161         3.819746e+07   
min          1.000000        3.000000        -1.951000e+08   
25%        439.000000      294.000000        -1.930000e+06   
50%       1143.000000      873.000000         1.500000e+05   
75%       1990.000000     2457.000000         7.600000e+06   
max       5070.000000    83678.000000         1.020000e+08   

       national_team_players  
count          123594.000000  
mean                6.551669  
std                 5.306323  
min                 0.000000  
25%                 2.000000  
50%                 6.000000  
75%                10.000000  
max                22.000000

Everything looks good, Lets get going with the preprocessing.

First, to avoid data leakage we will do our train test split before any proprocessing.

## Splitting

In [7]:
#Split my X and y
y=df["market_value"]
X=df.drop(["market_value"],axis=1)

In [8]:
#import splitter
from sklearn.model_selection import train_test_split

#split between train and val_test
X_train, X_val_test, y_train, y_val_test = train_test_split(X, y, test_size=0.33, random_state=100)

In [9]:
#split between test and val
X_val, X_test, y_val, y_test = train_test_split(X_val_test, y_val_test, test_size=0.5, random_state=100)

In [10]:
#check shape
X_train.shape

(82807, 21)

In [11]:
#check shape
X_val.shape

(20393, 21)

In [12]:
#check shape
X_test.shape

(20394, 21)

my splitting seems to have been correctly done. All the rows add up to the original size of the dataframe.

## Feature Encoding

Since I've got the month and year collumn i dont need the date collumn so I'll drop that.

In [13]:
X_train=X_train.drop(["date"],axis=1)
X_test=X_test.drop(["date"],axis=1)
X_val=X_val.drop(["date"],axis=1)

I need to encode the following collumns to be processed:

- `country_of_citizenship`
- `sub_position`
- `date`
- `position` 
- `foot` 
- `league_id`
  

We'll first look at the cardinality of all our categorical collumns to decide what type of encoding we should do.

In [14]:
df.nunique()

last_season_x                  12
country_of_citizenship        155
sub_position                   13
position                        4
foot                            3
height_in_cm                   48
market_value                  223
highest_ever_market_value     157
date                         2141
age                            30
contract_days_left             89
month                          12
year                           12
goals                          51
assists                        31
yellow_cards                   23
red_cards                       4
minutes_played               4079
player_club_id                426
league_id                      14
net_transfer_record           272
national_team_players          23
dtype: int64

the `foot`,`position`,`subposition` and `domestic competition id` collumns all have a low enough cardinality so one hot encoding would definitely be suitable.

In [15]:
#for train
encodecols = ["foot", "position", "sub_position", "league_id"]
X_train = pd.get_dummies(X_train, columns=encodecols, prefix=encodecols)

#for val
X_val=pd.get_dummies(X_val,columns=encodecols,prefix=encodecols)

#for test
X_test=pd.get_dummies(X_test,columns=encodecols,prefix=encodecols)

In [16]:
X_train.sample(5)

last_season_x country_of_citizenship  height_in_cm  \
27490            2023                 Russia         180.0   
105732           2020               Paraguay         184.0   
67585            2023            Netherlands         179.0   
103483           2020                Uruguay         185.0   
37763            2019                 Serbia         173.0   

        highest_ever_market_value  age  contract_days_left  month  year  \
27490                    700000.0   21                 232      7  2020   
105732                  5000000.0   25                 781      5  2021   
67585                   1000000.0   38                 232      7  2023   
103483                   300000.0   20                1327      1  2021   
37763                   1500000.0   29                -134      8  2020   

        goals  assists  yellow_cards  red_cards  minutes_played  \
27490       0        0             0          0             385   
105732      2        2             3          0            1004   
67585       0        0             1          0             495   
103483      0        0             1          0              80   
37763       0        2             1          0             602   

        player_club_id  net_transfer_record  national_team_players  foot_both  \
27490              121            7600000.0                     10          0   
105732           11282             800000.0                      5          0   
67585             1269            4750000.0                      0          0   
103483             940           29500000.0                      8          0   
37763              128                  0.0                      0          0   

        foot_left  foot_right  position_Attack  position_Defender  \
27490           0           1                1                  0   
105732          0           1                1                  0   
67585           0           1                0                  1   
103483          0           1                1                  0   
37763           0           1                0                  0   

        position_Goalkeeper  position_Midfield  \
27490                     0                  0   
105732                    0                  0   
67585                     0                  0   
103483                    0                  0   
37763                     0                  1   

        sub_position_Attacking Midfield  sub_position_Central Midfield  \
27490                                 0                              0   
105732                                0                              0   
67585                                 0                              0   
103483                                0                              0   
37763                                 0                              0   

        sub_position_Centre-Back  sub_position_Centre-Forward  \
27490                          0                            0   
105732                         0                            1   
67585                          1                            0   
103483                         0                            1   
37763                          0                            0   

        sub_position_Defensive Midfield  sub_position_Goalkeeper  \
27490                                 0                        0   
105732                                0                        0   
67585                                 0                        0   
103483                                0                        0   
37763                                 1                        0   

        sub_position_Left Midfield  sub_position_Left Winger  \
27490                            0                         0   
105732                           0                         0   
67585                            0                         0   
103483                           0                         0   
37763        

In [17]:
X_test.sample(5)

last_season_x country_of_citizenship  height_in_cm  \
71424            2020          Guinea-Bissau         184.0   
73645            2023                  Spain         182.0   
28682            2022             Guadeloupe         184.0   
114340           2017                 Sweden         181.0   
10530            2022                Croatia         186.0   

        highest_ever_market_value  age  contract_days_left  month  year  \
71424                   2500000.0   25                 232      6  2019   
73645                  10000000.0   26                 232      1  2015   
28682                    500000.0   20                 232      9  2022   
114340                  3500000.0   24                 202      2  2015   
10530                  18000000.0   24                 232     12  2018   

        goals  assists  yellow_cards  red_cards  minutes_played  \
71424       5        0             4          0            1281   
73645       2        6             9          0            4707   
28682       0        0             0          0              26   
114340      0        0             0          0              98   
10530       3        1             1          0            1064   

        player_club_id  net_transfer_record  national_team_players  foot_both  \
71424              855            3700000.0                      1          0   
73645              621                  0.0                      3          0   
28682              826            3640000.0                     12          0   
114340             931          -15870000.0                     14          0   
10530               33           10300000.0                      4          0   

        foot_left  foot_right  position_Attack  position_Defender  \
71424           0           1                1                  0   
73645           0           1                0                  1   
28682           0           1                1                  0   
114340          1           0                1                  0   
10530           0           1                1                  0   

        position_Goalkeeper  position_Midfield  \
71424                     0                  0   
73645                     0                  0   
28682                     0                  0   
114340                    0                  0   
10530                     0                  0   

        sub_position_Attacking Midfield  sub_position_Central Midfield  \
71424                                 0                              0   
73645                                 0                              0   
28682                                 0                              0   
114340                                0                              0   
10530                                 0                              0   

        sub_position_Centre-Back  sub_position_Centre-Forward  \
71424                          0                            1   
73645                          0                            0   
28682                          0                            1   
114340                         0                            0   
10530                          0                            0   

        sub_position_Defensive Midfield  sub_position_Goalkeeper  \
71424                                 0                        0   
73645                                 0                        0   
28682                                 0                        0   
114340                                0                        0   
10530                                 0                        0   

        sub_position_Left Midfield  sub_position_Left Winger  \
71424                            0                         0   
73645                            0                         0   
28682                            0                         0   
114340                           0                         0   
10530        

In [18]:
X_val.sample(5)

last_season_x country_of_citizenship  height_in_cm  \
31740            2023                England         188.0   
78849            2023         Czech Republic         182.0   
54642            2019               Portugal         184.0   
101379           2023                Ukraine         179.0   
15641            2023                Ukraine         185.0   

        highest_ever_market_value  age  contract_days_left  month  year  \
31740                  42000000.0   21                1327     10  2018   
78849                   8000000.0   19                1327      7  2016   
54642                    800000.0   23                -134      3  2013   
101379                   800000.0   27                  50     12  2022   
15641                   3500000.0   22                 689      5  2021   

        goals  assists  yellow_cards  red_cards  minutes_played  \
31740       0        0             4          0            2261   
78849       2        3             0          0             652   
54642       0        0             0          0             738   
101379      2        0             4          0            1116   
15641       2        0             4          0             971   

        player_club_id  net_transfer_record  national_team_players  foot_both  \
31740               31         -111300000.0                     16          0   
78849              610           47100000.0                     11          0   
54642              979           -2050000.0                      2          0   
101379           83678             300000.0                      0          0   
15641              338            1200000.0                      7          1   

        foot_left  foot_right  position_Attack  position_Defender  \
31740           0           1                0                  1   
78849           1           0                1                  0   
54642           0           1                0                  0   
101379          1           0                0                  1   
15641           0           0                0                  1   

        position_Goalkeeper  position_Midfield  \
31740                     0                  0   
78849                     0                  0   
54642                     1                  0   
101379                    0                  0   
15641                     0                  0   

        sub_position_Attacking Midfield  sub_position_Central Midfield  \
31740                                 0                              0   
78849                                 0                              0   
54642                                 0                              0   
101379                                0                              0   
15641                                 0                              0   

        sub_position_Centre-Back  sub_position_Centre-Forward  \
31740                          1                            0   
78849                          0                            0   
54642                          0                            0   
101379                         0                            0   
15641                          1                            0   

        sub_position_Defensive Midfield  sub_position_Goalkeeper  \
31740                                 0                        0   
78849                                 0                        0   
54642                                 0                        1   
101379                                0                        0   
15641                                 0                        0   

        sub_position_Left Midfield  sub_position_Left Winger  \
31740                            0                         0   
78849                            0                         0   
54642                            0                         0   
101379                           0                         0   
15641        

This seems to have worked well.The only collumn i need to deal with now is the `country_of_citizenship` collumn which consists of 155 unique values which on its own is far too many to one hot encode. The likely solution to this would be to get the top N-countries of citizenship in terms of numbers and have one designated collumn named other for every other nation other than them. This seems viable as nationalities in this dataset arent uniformly distributed as seen in our EDA in the previous notebook as the top 10 nationalities accounted for a disproportionate amount of the dataset.

In [19]:
#get the value counts for each nation
country_counts = df["country_of_citizenship"].value_counts()

#get the top 50 and its index
top_countries=country_counts.head(25).index



In [20]:
top_countries

Index(['Spain', 'France', 'Netherlands', 'Brazil', 'Turkey', 'Germany',
       'Portugal', 'Italy', 'Russia', 'England', 'Ukraine', 'Belgium',
       'Denmark', 'Greece', 'Argentina', 'Scotland', 'Serbia', 'Croatia',
       'Senegal', 'Nigeria', 'Sweden', 'Morocco', 'Uruguay', 'Cote d'Ivoire',
       'Poland'],
      dtype='object')

I need to now set everything not in that index to "Other".

In [21]:
#create lambda function setting "Other" if not in index for train,test and val
X_train["country_of_citizenship"]=X_train["country_of_citizenship"].apply(lambda x: x if x in top_countries else 'Other')
X_val["country_of_citizenship"]=X_val["country_of_citizenship"].apply(lambda x: x if x in top_countries else 'Other')
X_test["country_of_citizenship"]=X_test["country_of_citizenship"].apply(lambda x: x if x in top_countries else 'Other')

In [22]:
X_train.head()

last_season_x country_of_citizenship  height_in_cm  \
76575            2022                  Italy         184.0   
115759           2021                  Spain         180.0   
38683            2022                  Other         187.0   
44111            2023                  Spain         183.0   
90042            2022                  Other         181.0   

        highest_ever_market_value  age  contract_days_left  month  year  \
76575                  18000000.0   24                 597      6  2018   
115759                  5000000.0   20                -134      1  2015   
38683                   4500000.0   26                 232      6  2018   
44111                    800000.0   18                 597      2  2020   
90042                   3000000.0   22                1327     10  2017   

        goals  assists  yellow_cards  red_cards  minutes_played  \
76575       0        0             0          0              23   
115759      1        0             0          0             161   
38683       2        0             6          0             810   
44111       0        0             1          0             281   
90042       0        0             4          0            1211   

        player_club_id  net_transfer_record  national_team_players  foot_both  \
76575                5          -45000000.0                     15          0   
115759             150           43500000.0                      7          0   
38683              618           12850000.0                      3          0   
44111             1108           -6300000.0                      3          0   
90042              618           12850000.0                      3          0   

        foot_left  foot_right  position_Attack  position_Defender  \
76575           0           1                0                  1   
115759          0           1                1                  0   
38683           0           1                0                  0   
44111           1           0                0                  1   
90042           0           1                0                  1   

        position_Goalkeeper  position_Midfield  \
76575                     0                  0   
115759                    0                  0   
38683                     0                  1   
44111                     0                  0   
90042                     0                  0   

        sub_position_Attacking Midfield  sub_position_Central Midfield  \
76575                                 0                              0   
115759                                0                              0   
38683                                 0                              1   
44111                                 0                              0   
90042                                 0                              0   

        sub_position_Centre-Back  sub_position_Centre-Forward  \
76575                          0                            0   
115759                         0                            0   
38683                          0                            0   
44111                          0                            0   
90042                          0                            0   

        sub_position_Defensive Midfield  sub_position_Goalkeeper  \
76575                                 0                        0   
115759                                0                        0   
38683                                 0                        0   
44111                                 0                        0   
90042                                 0                        0   

        sub_position_Left Midfield  sub_position_Left Winger  \
76575                            0                         0   
115759                           0                         1   
38683                            0                         0   
44111                            0                         0   
90042        

I can now one hot encoding without excessively increasing dimensionality.

In [23]:
#train
X_train = pd.get_dummies(X_train, columns=["country_of_citizenship"], prefix="Country")

#val
X_val = pd.get_dummies(X_val, columns=["country_of_citizenship"], prefix="Country")

#test
X_test = pd.get_dummies(X_test, columns=["country_of_citizenship"], prefix="Country")


In [24]:
X_train.sample(5)

last_season_x  height_in_cm  highest_ever_market_value  age  \
109998           2021         178.0                   250000.0   26   
83253            2023         184.0                 15000000.0   33   
94279            2023         190.0                  2000000.0   27   
83954            2023         178.0                  1250000.0   31   
39299            2023         179.0                 22000000.0   23   

        contract_days_left  month  year  goals  assists  yellow_cards  \
109998                 232      5  2021      0        0             0   
83253                  232      7  2016      0        0             0   
94279                  597     12  2022      3        1             2   
83954                  232      6  2022      0        0             0   
39299                  232      2  2016     19        7             9   

        red_cards  minutes_played  player_club_id  net_transfer_record  \
109998          0              99           10484           -1000000.0   
83253           1            3831             131          102000000.0   
94279           0            1104            1390           -8600000.0   
83954           0            1530            1269            4750000.0   
39299           0            3404             610           47100000.0   

        national_team_players  foot_both  foot_left  foot_right  \
109998                      2          0          1           0   
83253                      15          0          0           1   
94279                       9          0          0           1   
83954                       0          0          1           0   
39299                      11          0          0           1   

        position_Attack  position_Defender  position_Goalkeeper  \
109998                1                  0                    0   
83253                 0                  0                    1   
94279                 0                  0                    0   
83954                 0                  0                    1   
39299                 0                  0                    0   

        position_Midfield  sub_position_Attacking Midfield  \
109998                  0                                0   
83253                   0                                0   
94279                   1                                0   
83954                   0                                0   
39299                   1                                0   

        sub_position_Central Midfield  sub_position_Centre-Back  \
109998                              0                         0   
83253                               0                         0   
94279                               1                         0   
83954                               0                         0   
39299                               1                         0   

        sub_position_Centre-Forward  sub_position_Defensive Midfield  \
109998                            0                                0   
83253                             0                                0   
94279                             0                                0   
83954                             0                                0   
39299                             0                                0   

        sub_position_Goalkeeper  sub_position_Left Midfield  \
109998                        0                           0   
83253                         1                           0   
94279                         0                           0   
83954                         1                           0   
39299                         0                           0   

        sub_position_Left Winger  sub_position_Left-Back  \
109998                         1                       0   
83253                          0                       0   
94279                          0                       0   
83954                          0                       0   
392

I've now completed my feature encoding and can move onto the next step of my preprocessing.

## Scaling Data

I'll use the StandardScaler as there will be quite a few outliers as bigger teams in better leagues have a disproportionate amount of funds to spend on players which min-max scaling is very sensitive to.

In [25]:
from sklearn.preprocessing import StandardScaler
#instantiate
scaler=StandardScaler()

#fit and transform
#train
X_train_s=scaler.fit_transform(X_train)

#val
X_val_s=scaler.transform(X_val)


#test
X_test_s=scaler.transform(X_test)




They're now in array form and I want it back in dataframe format.

In [26]:
#train
X_train_s = pd.DataFrame(X_train_s, columns=X_train.columns)

#val
X_val_s = pd.DataFrame(X_val_s, columns=X_val.columns)


#test
X_test_s = pd.DataFrame(X_test_s, columns=X_test.columns)



In [27]:
X_train_s.sample(10)

last_season_x  height_in_cm  highest_ever_market_value       age  \
54326       0.686577      1.158159                  -0.332696 -0.511092   
4125       -1.488296      0.709356                  -0.533108  0.240506   
73780       0.686577     -0.637050                   1.413748 -1.262690   
37161      -2.793220      1.307759                  -0.550286 -0.260559   
65092       0.686577      0.559756                  -0.447217 -0.010027   
81140       0.686577     -0.337849                  -0.361326 -0.260559   
77634       0.686577      1.606961                  -0.389957 -0.761625   
49314      -0.183373     -0.337849                  -0.304066  0.741572   
43474      -2.793220      0.410155                  -0.504478 -1.262690   
11864       0.686577     -0.188248                  -0.544560 -1.012157   

       contract_days_left     month      year     goals   assists  \
54326           -0.447889  0.439408 -2.245493 -0.262411 -0.612944   
4125            -1.223367 -0.118134 -0.566144 -0.532255 -0.205736   
73780            1.098831 -0.396905 -1.909624  0.007433  0.608680   
37161           -0.447889  0.439408 -0.902014 -0.532255 -0.612944   
65092           -0.447889 -1.511990  0.105595  0.007433 -0.205736   
81140            1.098831 -1.511990  0.777335 -0.532255 -0.612944   
77634            1.872191  0.996950  0.441465 -0.532255 -0.612944   
49314           -1.286931  0.996950 -0.566144 -0.532255  0.201472   
43474           -0.833509  0.160637 -2.245493 -0.262411  0.201472   
11864            1.098831 -0.954448  1.113205 -0.262411 -0.205736   

       yellow_cards  red_cards  minutes_played  player_club_id  \
54326      0.094172  -0.258501       -1.048029        0.125582   
4125       2.906462  -0.258501       -0.084078       -0.243821   
73780      0.445709   3.500180        1.069096       -0.264571   
37161     -0.257364  -0.258501       -1.173848       -0.206804   
65092     -0.960436  -0.258501       -0.755773       -0.093191   
81140     -0.257364  -0.258501       -0.138567       -0.358715   
77634     -0.608900  -0.258501       -0.136585       -0.215258   
49314      0.445709  -0.258501        1.557511       -0.258167   
43474     -0.257364  -0.258501        0.737212        0.035665   
11864     -0.608900  -0.258501       -0.900415       -0.284809   

       net_transfer_record  national_team_players  foot_both  foot_left  \
54326            -0.002528              -1.239765  -0.203311  -0.577466   
4125              0.023622              -0.673275  -0.203311  -0.577466   
73780             0.156986              -0.106785  -0.203311   1.731702   
37161             0.065462              -1.239765  -0.203311  -0.577466   
65092             0.422669              -0.862105  -0.203311   1.731702   
81140             0.601534               0.459705  -0.203311  -0.577466   
77634             0.039312              -1.239765  -0.203311  -0.577466   
49314            -2.807100               1.403856  -0.203311  -0.577466   
43474            -0.002528              -0.862105  -0.203311   1.731702   
11864            -0.012988              -0.295615  -0.203311  -0.577466   

       foot_right  position_Attack  position_Defender  position_Goalkeeper  \
54326    0.638745        -0.635055           1.407452            -0.297908   
4125     0.638745        -0.635055          -0.710504            -0.297908   
73780   -1.565570        -0.635055           1.407452            -0.297908   
37161    0.638745        -0.635055          -0.710504            -0.297908   
65092   -1.565570        -0.635055          -0.710504            -0.297908   
81140    0.638745        -0.635055          -0.710504            -0.297908   
77634    0.638745        -0.635055          -0.710504             3.356740   
49314    0.638745        -0.635055           1.407452            -0.297908   
43474   -1.565570        -0.635055           1.407452            -0.297908   
11864    0.638745         1.574668          -0.710504            -0.297908   

       po

In [28]:
#single out Cristiano Ronaldo using scaled data
X_train_s.loc[
    (X_train_s['position_Attack'] > 1.0) & (X_train_s['position_Attack'] < 2.0) &
    (X_train_s['Country_Portugal'] > 4.0) & (X_train_s['Country_Portugal'] < 5.0) &
    (X_train_s['foot_right'] > -1.6) & (X_train_s['foot_right'] < -1.5)&
    (X_train_s["height_in_cm"]>0.7) & (X_train_s["height_in_cm"]<0.71)
]



last_season_x  height_in_cm  highest_ever_market_value       age  \
1122        0.686577      0.709356                   0.268539 -0.761625   
3336        0.686577      0.709356                   0.268539 -0.010027   
10245       0.686577      0.709356                   0.268539 -0.761625   
14562       0.686577      0.709356                   0.268539 -0.010027   
26416       0.686577      0.709356                   0.268539  1.493170   
27580       0.686577      0.709356                   0.268539  0.491039   
40603       0.686577      0.709356                   0.268539  0.741572   
55824       0.686577      0.709356                   0.268539  0.992104   
58852       0.686577      0.709356                   0.268539  1.242637   
69160       0.686577      0.709356                   0.268539  1.242637   
76847       0.686577      0.709356                   0.268539  0.741572   
80396       0.686577      0.709356                   0.268539  0.240506   
82494       0.686577      0.709356                   0.268539  0.491039   

       contract_days_left     month      year     goals   assists  \
1122             1.098831  0.160637 -1.573754 -0.262411 -0.612944   
3336             1.098831 -0.118134 -0.566144  1.356653  0.608680   
10245            1.098831 -1.233219 -1.237884 -0.532255 -0.612944   
14562            1.098831  1.275721 -0.566144  1.356653  0.608680   
26416            1.098831 -0.118134  1.449075  2.436029  2.237513   
27580            1.098831 -0.118134  0.105595  3.515405  0.608680   
40603            1.098831  0.439408  0.441465  5.134469  3.459137   
55824            1.098831 -0.396905  0.777335  2.705873  2.644721   
58852            1.098831 -0.396905  1.113205  3.245561  2.644721   
69160            1.098831 -0.954448  1.449075  2.436029  2.237513   
76847            1.098831 -1.511990  0.777335  2.705873  2.644721   
80396            1.098831 -0.118134 -0.230274  3.515405  1.423097   
82494            1.098831 -0.675677  0.441465  5.134469  3.459137   

       yellow_cards  red_cards  minutes_played  player_club_id  \
1122       0.445709  -0.258501       -0.669582       -0.092679   
3336       0.094172  -0.258501        0.156662       -0.265468   
10245     -0.608900  -0.258501       -1.047038       -0.092679   
14562      0.094172  -0.258501        0.156662       -0.265468   
26416      0.797245  -0.258501        0.490527       -0.360124   
27580      0.797245  -0.258501        0.956156       -0.265468   
40603      1.500317  -0.258501        1.822028       -0.265468   
55824      2.203390  -0.258501        1.802214       -0.265468   
58852      1.851853  -0.258501        1.121603       -0.360124   
69160      0.797245  -0.258501        0.490527       -0.360124   
76847      2.203390  -0.258501        1.802214       -0.265468   
80396      0.094172  -0.258501        0.811514       -0.265468   
82494      1.500317  -0.258501        1.822028       -0.265468   

       net_transfer_record  national_team_players  foot_both  foot_left  \
1122              0.204056              -0.673275  -0.203311   1.731702   
3336             -0.177732              -0.106785  -0.203311   1.731702   
10245             0.204056              -0.673275  -0.203311   1.731702   
14562            -0.177732              -0.106785  -0.203311   1.731702   
26416             1.814888               0.270875  -0.203311   1.731702   
27580            -0.177732              -0.106785  -0.203311   1.731702   
40603            -0.177732              -0.106785  -0.203311   1.731702   
55824            -0.177732              -0.106785  -0.203311   1.731702   
58852             1.814888               0.270875  -0.203311   1.731702   
69160             1.814888               0.270875  -0.203311   1.731702   
76847            -0.177732              -0.106785  -0.203311   1.731702   
80396            -0.177732              -0.106785  -0.203311   1.731702   
82494            -0.177732              -0.106785  -0.203311   1.731702   

       foot_r

Data looks correctly scaled. We can now move onto our Baseline linear Model.

### Data Export

In [29]:
import pickle
df.to_pickle("data1/df.pkl")
X_train_s.to_pickle('data1/X_train_s.pkl')
X_val_s.to_pickle('data1/X_val_s.pkl')
X_test_s.to_pickle('data1/X_test_s.pkl')
y_train.to_pickle('data1/y_train.pkl')
y_val.to_pickle('data1/y_val.pkl')
y_test.to_pickle('data1/y_test.pkl')